plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -649,6 +695,82 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
+      return eosio::chain::string_to_name(n); // n with invalid chars are encoded as 0 without throwing exception
+   }
+
+   // called from multiple threads
+   void on_vote(uint32_t connection_id, vote_result_t status, const vote_message_ptr& msg,
+                const finalizer_authority_ptr& active_auth, const finalizer_authority_ptr& pending_auth) {
+      switch( status ) {
+      case vote_result_t::success:
+      case vote_result_t::duplicate:
+         break;
+      case vote_result_t::unknown_public_key:
+      case vote_result_t::invalid_signature:
+      case vote_result_t::max_exceeded:
+      case vote_result_t::unknown_block:
+         return;
+      default:
+         assert(false); // should never happen
+      }
+
+      if (!active_auth && !pending_auth) {
+         elog("vote signal contains no valid authority ${m}", ("m", msg));
+         return;
+      }
+
+      auto now = fc::time_point::now();
+      account_name auth_desc = active_auth ? to_account_name_safe(active_auth->description)
+                                           : to_account_name_safe(pending_auth->description);
+      if (auth_desc.empty()) {
+         ilog("Finalizer authority description is not a valid producer name ${d}",
+              ("d", active_auth ? active_auth->description : pending_auth->description));
+         // running with core contract that does not associate finalizer_authority->description with a producer
+         // reset times otherwise the producer would pause
+         _last_other_vote_received.store(now, std::memory_order_relaxed);
+         _last_producer_vote_received.store(now, std::memory_order_relaxed);
+         return;
+      }
+
+      if (_producers.contains(auth_desc)) { // _producers not modified, thread safe
+         _last_producer_vote_received.store(now, std::memory_order_relaxed);
+      } else {
+         _last_other_vote_received.store(now, std::memory_order_relaxed);
+      }
+   }
+
+   // true if paused due to not receiving votes
+   bool is_implicitly_paused() const {
+      if (_producers.empty() || _production_pause_vote_timeout.count() == 0)
+         return false;
+      if (!_is_savanna_active)
+         return false; // no implicit pause in legacy
+      if (_producers.contains(eosio::chain::config::system_account_name)) // disable implicit pause for eosio
+         return false;
+
+      auto last_producer_vote_received = _last_producer_vote_received.load(std::memory_order_relaxed);
+      auto last_other_vote_received    = _last_other_vote_received.load(std::memory_order_relaxed);
+
+      // need a vote within timeout of last accepted block
+      return (_is_producer_active_finalizer &&
+              _accepted_block_time - last_producer_vote_received > _production_pause_vote_timeout)
+          || (_other_active_finalizers &&
+              _accepted_block_time - last_other_vote_received    > _production_pause_vote_timeout);
```

Comment:
Not an issue, but I liked the previous version better. Here we can have negative time values and I think it is less clear.